hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -76,18 +82,41 @@ public JobExecutionResult execute() throws Exception {
         DataStream<TaxiFare> fares = env.addSource(source);
 
         // replace this with your solution
-        if (true) {
-            throw new MissingSolutionException();
+        fares.assignTimestampsAndWatermarks(WatermarkStrategy.<TaxiFare>forMonotonousTimestamps()
+                                                             .withTimestampAssigner((fare, timestamp) -> fare.getEventTimeMillis()))
+             .keyBy(fare -> fare.driverId)
+             .window(TumblingEventTimeWindows.of(Time.hours(1)))
+             .reduce(new HourlyTipsAccumulator(), new HourlyTipsProcessor())
+             .forward()
```

Comment:
I don't see any reason to use `.forward()` because you're passing it `keyBy()` function and you stream will be repartitioned anyway, so Flink will ignore your  `.forward()` instruction
pls see you topology

<img width="966" alt="Screen Shot 2024-05-24 at 09 03 59" src="https://github.com/gerson23/flink-training/assets/153856856/3ce84376-8059-4ae1-9a7d-ab9fb5b027f9">

or you could print out execution plan:
`env.getExecutionPlan()` and see how it looks like:
```
{
  "nodes" : [ {
    "id" : 1,
    "type" : "Source: Custom Source",
    "pact" : "Data Source",
    "contents" : "Source: Custom Source",
    "parallelism" : 1
  }, {
    "id" : 2,
    "type" : "Timestamps/Watermarks",
    "pact" : "Operator",
    "contents" : "Timestamps/Watermarks",
    "parallelism" : 1,
    "predecessors" : [ {
      "id" : 1,
      "ship_strategy" : "FORWARD",
      "side" : "second"
    } ]
  }, {
    "id" : 4,
    "type" : "TumblingEventTimeWindows",
    "pact" : "Operator",
    "contents" : "Window(TumblingEventTimeWindows(3600000), EventTimeTrigger, HourlyTipsAccumulator, HourlyTipsProcessor)",
    "parallelism" : 8,
    "predecessors" : [ {
      "id" : 2,
      "ship_strategy" : "HASH",
      "side" : "second"
    } ]
  }, {
    "id" : 7,
    "type" : "TumblingEventTimeWindows",
    "pact" : "Operator",
    "contents" : "Window(TumblingEventTimeWindows(3600000), EventTimeTrigger, ComparableAggregator, PassThroughWindowFunction)",
    "parallelism" : 8,
    "predecessors" : [ {
      "id" : 4,
      "ship_strategy" : "HASH",
      "side" : "second"
    } ]
  }, {
    "id" : 8,
    "type" : "Sink: Unnamed",
    "pact" : "Data Sink",
    "contents" : "Sink: Unnamed",
    "parallelism" : 8,
    "predecessors" : [ {
      "id" : 7,
      "ship_strategy" : "FORWARD",
      "side" : "second"
    } ]
  } ]
}
```

it will be the same with or without `.forward()` instruction

Reply:
Yeah, I see it now. Removed the `.forward` instruction, thanks.